File: src/dotnet/PolyBus/Transport/InMemory/InMemoryEndpoint.cs
using System.Collections.Concurrent;
using PolyBus.Transport.Transactions;
using PolyBus.Transport.Transactions.Messages;

namespace PolyBus.Transport.InMemory;

/// <summary>
/// An implementation of an in-memory transport endpoint.
/// </summary>
public class InMemoryEndpoint(InMemoryMessageBroker broker, IPolyBus bus) : ITransport
{
    /// <summary>
    /// The associated PolyBus instance.
    /// </summary>
    public IPolyBus Bus => bus;

    public Action<IncomingMessage>? DeadLetterHandler { get; set; }

    public bool Active { get; private set; }

    public string DeadLetterEndpoint => $"{bus.Name}.dead.letters";

    /// <summary>
    /// If active, handles an incoming message by creating a transaction and executing the handlers for the bus.
    /// </summary>
    public async Task HandleMessage(IncomingMessage message, bool isDeadLetter)
    {
        if (Active)
        {
            if (isDeadLetter)
            {
                DeadLetterHandler?.Invoke(message);
            }
            else
            {
                var transaction = await bus.CreateIncomingTransaction(message);

                await bus.Send(transaction);
            }
        }
    }

    public Task Handle(Transaction transaction)
    {
        if (!Active)
        {
            throw new PolyBusNotStartedError();
        }

        broker.Send(transaction);

        return Task.CompletedTask;
    }
    public bool SupportsDelayedCommands => true;
    public bool SupportsCommandMessages => true;

    public Task Subscribe(MessageInfo messageInfo)
    {
        if (!Active)
        {
            throw new PolyBusNotStartedError();
        }

        _subscriptions[messageInfo.ToString(false)] = true;

        return Task.CompletedTask;
    }
    public bool IsSubscribed(MessageInfo messageInfo) => _subscriptions.ContainsKey(messageInfo.ToString(false));
    public bool SupportsSubscriptions => true;
    readonly ConcurrentDictionary<string, bool> _subscriptions = new();

    public Task Start()
    {
        Active = true;
        return Task.CompletedTask;
    }

    public Task Stop()
    {
        Active = false;
        return Task.CompletedTask;
    }
}

File: src/kotlin/src/main/kotlin/polybus/transport/inmemory/InMemoryEndpoint.kt
package polybus.transport.inmemory

import polybus.IPolyBus
import polybus.transport.ITransport
import polybus.transport.PolyBusNotStartedError
import polybus.transport.transactions.Transaction
import polybus.transport.transactions.messages.IncomingMessage
import polybus.transport.transactions.messages.MessageInfoValue
import java.util.concurrent.ConcurrentHashMap

open class InMemoryEndpoint(
    private val _broker: InMemoryMessageBroker,
    private val _bus: IPolyBus
) : ITransport {
    val bus: IPolyBus
        get() = _bus

    var deadLetterHandler: ((IncomingMessage) -> Unit)? = null

    var active: Boolean = false
        private set

    override val deadLetterEndpoint: String
        get() = "${_bus.name}.dead.letters"

    open suspend fun handleMessage(message: IncomingMessage, isDeadLetter: Boolean) {
        if (!active) {
            return
        }
        if (isDeadLetter) {
            deadLetterHandler?.invoke(message)
            return
        }
        val transaction = _bus.createIncomingTransaction(message)
        _bus.send(transaction)
    }

    override suspend fun handle(transaction: Transaction) {
        if (!active) {
            throw PolyBusNotStartedError()
        }
        _broker.send(transaction)
    }

    override val supportsDelayedCommands: Boolean = true
    override val supportsCommandMessages: Boolean = true
    override val supportsSubscriptions: Boolean = true

    private val _subscriptions = ConcurrentHashMap<String, Boolean>()

    override suspend fun subscribe(messageInfo: MessageInfoValue) {
        if (!active) {
            throw PolyBusNotStartedError()
        }
        _subscriptions[messageInfo.toHeaderString(false)] = true
    }

    fun isSubscribed(messageInfo: MessageInfoValue): Boolean =
        _subscriptions.containsKey(messageInfo.toHeaderString(false))

    override suspend fun start() {
        active = true
    }

    override suspend fun stop() {
        active = false
    }
}
